<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · InformMe.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">InformMe.jl</span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Functions-1"><span>Functions</span></a></li><li><a class="tocitem" href="#Index-1"><span>Index</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/master/docs/src/index.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="InformMe.jl-Documentation-1"><a class="docs-heading-anchor" href="#InformMe.jl-Documentation-1">InformMe.jl Documentation</a><a class="docs-heading-anchor-permalink" href="#InformMe.jl-Documentation-1" title="Permalink"></a></h1><ul><li><a href="#InformMe.jl-Documentation-1">InformMe.jl Documentation</a></li><ul><li><a href="#Functions-1">Functions</a></li><li><a href="#Index-1">Index</a></li></ul></ul><h2 id="Functions-1"><a class="docs-heading-anchor" href="#Functions-1">Functions</a><a class="docs-heading-anchor-permalink" href="#Functions-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="InformMe.fastaToCpG-Tuple{Any}" href="#InformMe.fastaToCpG-Tuple{Any}"><code>InformMe.fastaToCpG</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function is used to analyze a reference genome in order to find  and store the locations of all CpG sites within each chromosome and  to compute the CpG densities at each CpG site as well as the distances  between neighboring CpG sites. A 1-based coordinate system is used,  in which the first base is assigned to position 1 and the location  of a CpG site is defined by the position of the C nucleotide on the  forward strand of the reference genome.</p><p>This function must be run ONLY ONCE before proceeding with analysis  of BAM files.</p><p>USAGE (default):</p><p>fastaToCpg(FASTAfilename)</p><p>USAGE (optional):</p><p>Example of optional usage with additional input parameters.  FastaToCpG(FASTAfilename,&#39;maxChrNum&#39;,23)</p><p>MADATORY INPUT:</p><p>FASTAfilename                Full path of FASTA-formatted reference genome to which                available BAM files have been aligned to.</p><p>OPTIONAL INPUTS:</p><p>outdir                Path where the output will be stored at.          Default: &quot;REFGENEDIR&quot;</p><p>maxChrNum                Maximum number of chromosomes to be processed. The                function will process chromosomes 1,2,...,maxChrNum.                Default value: 22</p><p>wsize                Window size used in CpG density calculations.                Default value: 1000</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/7529714c80924774974ff35e8f1bc9e70921eadb/src/fastaToCpG.jl#LL20-L62">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.convertBAMtoBits-Tuple{Any,Any}" href="#InformMe.convertBAMtoBits-Tuple{Any,Any}"><code>InformMe.convertBAMtoBits</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Runs the entire MethToBits pipeline (assuming single BAM file as input).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/7529714c80924774974ff35e8f1bc9e70921eadb/src/convertBAMtoBits.jl#LL20-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.matrixFromBam-Tuple{Any,Any}" href="#InformMe.matrixFromBam-Tuple{Any,Any}"><code>InformMe.matrixFromBam</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function processes a BAM file with aligned reads to a reference  genome and produces methylation information for nonoverlapping genomic  regions (containing the same number of base pairs) in a given chromosome.  The final output for each genomic region is a matrix with -1,0,1 values.  Each row of the matrix is a methylation read, whereas each column  represents a CpG site within the genomic region. A value of -1 indicates  no methylation information is available for the CPG site, 0 indicates  that the CpG site is unmethylated, and 1 indicates that the CpG site  is methylated.</p><p>This function depends on a working instalation of SAMtools that is on  the system path PATH.</p><p>Before running this function, FastaToCpG.m must be run ONCE.</p><p>USAGE (default):</p><p>matrixFromBam(bam<em>prefix,chr</em>num)</p><p>USAGE (optional):</p><p>Example of optional usage with additional input parameters.  matrixFromBam(bam<em>prefix,chr</em>num,&#39;reference_path&#39;,&#39;/path/to/ref&#39;)</p><p>MADATORY INPUTS:</p><p>bam<em>prefix                 Prefix of the BAM file (without the .bam extension). This                 file must be sorted from the least to the greatest base                 pair position along the reference sequence and must be                 indexed (i.e., the associated BAI file must be available).                 The file name must not contain &quot;.&quot; characters, but can                 contain &quot;</em>&quot; instead. Moreover, the file name should be                 unique.</p><p>chr_num                 Number representing the chromosome to be processed.</p><p>OPTIONAL INPUTS:</p><p>reference_path                 Path to the root subdirectory where the outputs of this                 function are stored.                 Default value: &quot;REFGENEDIR&quot;</p><p>bamfile_path                 Path to the subdirectory where the BAM file is located.                 Default value: &quot;BAMDIR&quot;</p><p>matrices_path                 Path to the subdirectory where the output of this function                 is stored.                 Default value: &quot;INTERMEDIATE&quot;</p><p>pairedEnds                 Flag for paired end read support. A value of 1 indicates                 that the sequencer employed paired end reads, whereas a                 value of 0 indicates that the sequencer employed single                 end reads.                 Default value: 1</p><p>numBasesToTrim                 A vector of integers specifying how many bases should be                 trimmed from the begining of each read. If the vector                 contains two integers, then the first integer specifies                 how many bases to trim from the first read in a read pair,                 whereas the second integer specifies how many bases should                 be trimmed from the second read in the pair. If the                 vector contains one integer, then all reads will have                 that number of bases trimmed from the beginning of the                 read. If no bases are to be trimmed, then this input                 must be set to 0.                 Default value: 0</p><p>regionSize                 The size of the genomic regions for which methylation                 information is produced (in number of base pairs).                 Default value: 3000</p><p>minCpGsReqToModel                 The minimum number of CpG sites within a genomic region                 required to perform statistical estimation.                 Default value: 10</p><p>The default values of regionSize and minCpGsReqToModel should only be  changed by an expert with a detailed understanding of the code and the  methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/7529714c80924774974ff35e8f1bc9e70921eadb/src/matrixFromBam.jl#LL21-L109">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.estParamsForChr-NTuple{5,Any}" href="#InformMe.estParamsForChr-NTuple{5,Any}"><code>InformMe.estParamsForChr</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function takes a list of BAM files (which correspond to the same  phenotype) and performs statistical model estimation within a specific  chromosome of interest. The function can be used on a computing cluster  to break the work of model estimation to many independent parallel job  processes. This is performed only after matrixFromBam.m.</p><p>USAGE (default):</p><p>estParamsForChr(mat<em>files,prefix,matrices</em>path,reference<em>path,chr</em>num)</p><p>USAGE (optional):</p><p>Example of optional usage with additional input parameters.  estParamsForChr(mat<em>files,prefix,matrices</em>path,reference<em>path,chr</em>num,...           &#39;regionSize&#39;,2000)</p><p>MANDATORY INPUTS:</p><p>mat_files         All the .mat files to be included in the model. This can be a         single .mat file or multiple files in the form of a comma-sepa-         rated list of files.</p><p>prefix         A string that specifies the name of the modeled phenotype.         The output files produced will contain this prefix.</p><p>matrices_path         A string that specifies the path to the directory that         where the output will be stored.</p><p>reference_path         A string that specifies the path to the directory that         contains the results of analysis of the reference genome         performed by FastaToCpG.m as well as the results of         methylation calling performed by matrixFromBam.m.</p><p>chr_num         Chromosome number 1 to 22 (in humans) specifying the         chromosome for which statistical estimation must be         performed.</p><p>OPTIONAL INPUTS:</p><p>regionSize         The size of the genomic region used for parameter         estimation (in number of base pairs).         Default value: 3000</p><p>boundaryConditions         Flag to decide if boundary conditions should be estimated         freely in MLE.         Default value: false</p><p>The default value of regionSize should only be changed by an expert with  a detailed understanding of the code and the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/7529714c80924774974ff35e8f1bc9e70921eadb/src/estParamsForChr.jl#LL21-L78">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.methAnalysisForChr-NTuple{4,Any}" href="#InformMe.methAnalysisForChr-NTuple{4,Any}"><code>InformMe.methAnalysisForChr</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function performs methylation analysis of a given chromosome in a  single phenotype. The function can be used on a computing cluster to  break the analysis work to many independent parallel job processes.  This is performed only after estParamsForChr.m in the Modeling  subdirectory is run to build the Ising models for the phenotype.</p><p>USAGE (default):</p><p>methAnalysisForChr(prefix,chr<em>num,reference</em>path,estimation_path)</p><p>USAGE (optional):</p><p>Example of optional usage with additional input parameters.  methAnalysisForChr(prefix,chr<em>num,reference</em>path,estimation_path,               &#39;outdir&#39;,&#39;/path/to/output&#39;)</p><p>MANDATORY INPUTS:</p><p>prefix                A string that specifies the name of the phenotype to be                analyzed.</p><p>chr_num                Chromosome number (1 to 22 in humans) specifying the                chromosome for which methylation analyis must be                performed.</p><p>reference_path                A string that specifies the path to the directory that                contains the results of analysis of the reference genome                performed by FastaToCpG.m as well as the results of                methylation calling performed by matrixFromBam.jl.</p><p>estimation_path                A string that specifies the path to the directory that                contains the results of parameter estimation performed                by estParamsForChr.m.</p><p>OPTIONAL INPUTS:</p><p>outdir                A string that specifies the path of the directory in which                the methylation analysis results are written.                Default value &#39;./results/&#39;</p><p>MSIflag                Flag that determines whether this function performs                computation of the methylation sensitivity index (MSI).                false: no MSI computation.                true: allow MSI computation.                Default value: false</p><p>ESIflag                Flag that determines whether this function performs                computation of the entropic sensitivity index (ESI).                false: no ESI computation.                true: allow ESI computation.                Default value: false</p><p>MCflag                Flag that determines whether this function performs                computation of turnover ratios, CpG entropies, capacities,                and relative dissipated energies of methylation                channels (MCs).                false: no MC computations.                true: allow MC computations.                Default value: false</p><p>regionSize                The size of the genomic regions used for parameter                estimation (in number of base pairs).                Default value: 3000</p><p>subregionSize                The size of the subregions of a genomic region used                for methylation analysis (in number of base pairs).                The ratio regionSize/subregionSize must be an integer.                Default value: 150</p><p>The default values of regionSize and subregionSize should only be  changed by an expert with a detailed understanding of the code and  the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/7529714c80924774974ff35e8f1bc9e70921eadb/src/methAnalysisForChr.jl#LL21-L105">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}" href="#InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}"><code>InformMe.makeBedsForMethAnalysis</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function makes BED files for the methylation analysis results  obtained by means of MethAnalysisForChr.m for a single phenotype.</p><p>USAGE (default):</p><p>makeBedsForMethAnalysis(prefix,analysis<em>path,reference</em>path)</p><p>USAGE (optional):</p><p>Example of optional usage with additional input parameters.  makeBedsForMethAnalysis(prefix,analysis<em>path,reference</em>path,&#39;outdir&#39;,     &#39;/path/to/output&#39;)</p><p>MANDATORY INPUTS:</p><p>prefix                A string that specifies the name of the phenotype.</p><p>analysis_path                A string that specifies the path of the directory in which                the model was constructed.         Default: &quot;INTERMEDIATE&quot;</p><p>reference_path                A string that specifies the path to the directory that                contains the results of analysis of the reference genome                performed by FastaToCpG.m as well as the results of                methylation calling performed by matrixFromBam.jl.         Default: &quot;REFGENEDIR&quot;</p><p>OPTIONAL INPUTS:</p><p>minChrNum                A number specifying the starting chromosome that will be                included in the BED files.                Default value: 1</p><p>maxChrNum                A number specifying the last chromosome that will be                included in the output BED files. Must be                maxChrNum &gt;= minChrNum.                Default value: 22</p><p>outdir                A string that specifies the path of the directory in which                the output BED files are written.                Default value &#39;./&#39;</p><p>MSIflag                Flag that determines whether this function performs                computation of the methylation sensitivity index (MSI).                false: no MSI computation.                true: allow MSI computation.                Default value: false</p><p>ESIflag                Flag that determines whether this function performs                computation of the entropic sensitivity index (ESI).                false: no ESI computation.                true: allow ESI computation.                Default value: false</p><p>MCflag                Flag that determines whether this function performs                computation of turnover ratios, CpG entropies, capacities,                and relative dissipated energies of methylation                channels (MCs).                false: no MC computations.                true: allow MC computations.                Default value: false</p><p>thresh                A scalar used as a threshold in methylation-based                classification.                Default value: 0.4</p><p>regionSize                The size of the genomic regions used for parameter                estimation (in number of base pairs).                Default value: 3000</p><p>subregionSize                The size of the subregions of a genomic region used                for methylation analysis (in number of base pairs).                The ratio regionSize/subregionSize must be an integer.                Default value: 150</p><p>The default values of thresh, regionSize, and subregionSize should only  be changed by an expert with a detailed understanding of the code and  the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/7529714c80924774974ff35e8f1bc9e70921eadb/src/makeBedsForMethAnalysis.jl#LL20-L112">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.diffMethAnalysisToBed" href="#InformMe.diffMethAnalysisToBed"><code>InformMe.diffMethAnalysisToBed</code></a> — <span class="docstring-category">Function</span></header><section><div><p>This function makes BED files for the differential version of the  methylation analysis results obtained by means of MethAnalysisForChr.m  applied on two dinstict phenotypes.</p><p>USAGE (default):</p><p>makeBedsForDiffMethAnalysis(prefix<em>1,prefix</em>2,analysis<em>path</em>1,...     analysis<em>path</em>2,reference_path)</p><p>MANDATORY INPUTS:</p><p>prefix_X                Strings with the first string specifying                the name of the first phenotype and the second string                specifying the name of the second phenotype used for                differential methylation analysis. Both phenotypes                must have already been analyzed with methAnalysisForChr.m.</p><p>analysis<em>path</em>X                A string that specifies the path of the directory in which                the methylation analysis results obtained by                MethAnalysisForChr.m are stored.         Default: &quot;INTERMEDIATE&quot;</p><p>reference_path                A string that specifies the path to the directory that                contains the results of analysis of the reference genome                performed by FastaToCpG.m as well as the results of                methylation calling performed by matrixFromBam.m.         Default: &quot;REFGENEDIR&quot;</p><p>OPTIONAL INPUTS:</p><p>minChrNum                A number specifying the starting chromosome that will be                included in the BED files.                Default value: 1</p><p>maxChrNum                A number specifying the last chromosome that will be                included in the outut BED files. Must be                maxChrNum &gt;= minChrNum.                Default value: 22</p><p>outdir                A string that specifies the path of the directory in which                the output BED files are written.                Default value &quot;INTERMEDIATE&quot;</p><p>MSIflag                Flag that determines whether this function performs                computation of the methylation sensitivity index (MSI).                0: no MSI computation.                1: allow MSI computation.                Default value: 0</p><p>ESIflag                Flag that determines whether this function performs                computation of the entropic sensitivity index (ESI).                0: no ESI computation.                1: allow ESI computation.                Default value: 0</p><p>MCflag                Flag that determines whether this function performs                computation of turnover ratios, CpG entropies, capacities,                and relative dissipated energies of methylation                channels (MCs).                0: no MC computations.                1: allow MC computations.                Default value: 0</p><p>regionSize                The size of the genomic regions used for parameter                estimation (in number of base pairs).                Default value: 3000</p><p>subregionSize                The size of the subregions of a genomic region used                for methylation analysis (in number of base pairs).                The ratio regionSize/subregionSize must be an integer.                Default value: 150</p><p>minNumCpG     The minimum number of CpG sites within an analysis                subregion required for performing full methylation-based                differential classification.                Default value: 2</p><p>thresh</p><pre><code class="language-none">           A scalar used as a threshold in methylation-based
           differential classification.
           Default value: 0.55</code></pre><p>threshDMU                A 1x6 vector containing threshold values used for                methylation-based differential classification.                Default value: [-1,-0.55,-0.1,0.1,0.55,1]</p><p>threshDEU                A 1x8 vector containing threshold values used for                entropy-based differential classification.                Default value: [-1,-0.5,-0.3,-0.05,0.05,0.3,0.5,1]</p><p>The default values of regionSize, subregionSize, minNumCpG, thresh,  threshDMU, and threshDEU should only be changed by an expert with  a detailed understanding of the code and the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/7529714c80924774974ff35e8f1bc9e70921eadb/src/diffMethAnalysisToBed.jl#LL20-L129">source</a></section></article><h2 id="Index-1"><a class="docs-heading-anchor" href="#Index-1">Index</a><a class="docs-heading-anchor-permalink" href="#Index-1" title="Permalink"></a></h2><ul><li><a href="#InformMe.convertBAMtoBits-Tuple{Any,Any}"><code>InformMe.convertBAMtoBits</code></a></li><li><a href="#InformMe.diffMethAnalysisToBed"><code>InformMe.diffMethAnalysisToBed</code></a></li><li><a href="#InformMe.estParamsForChr-NTuple{5,Any}"><code>InformMe.estParamsForChr</code></a></li><li><a href="#InformMe.fastaToCpG-Tuple{Any}"><code>InformMe.fastaToCpG</code></a></li><li><a href="#InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}"><code>InformMe.makeBedsForMethAnalysis</code></a></li><li><a href="#InformMe.matrixFromBam-Tuple{Any,Any}"><code>InformMe.matrixFromBam</code></a></li><li><a href="#InformMe.methAnalysisForChr-NTuple{4,Any}"><code>InformMe.methAnalysisForChr</code></a></li></ul></article></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 2 December 2019 03:28">Monday 2 December 2019</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
