var documenterSearchIndex = {"docs":
[{"location":"#InformMe.jl-Documentation-1","page":"Home","title":"InformMe.jl Documentation","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"","category":"page"},{"location":"#Functions-1","page":"Home","title":"Functions","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"fastaToCpG(FASTAfilename;\n                    outdir=\"./\",\n                    wsize=1000)","category":"page"},{"location":"#InformMe.fastaToCpG-Tuple{Any}","page":"Home","title":"InformMe.fastaToCpG","text":"This function is used to analyze a reference genome in order to find  and store the locations of all CpG sites within each chromosome and  to compute the CpG densities at each CpG site as well as the distances  between neighboring CpG sites. A 1-based coordinate system is used,  in which the first base is assigned to position 1 and the location  of a CpG site is defined by the position of the C nucleotide on the  forward strand of the reference genome.\n\nThis function must be run ONLY ONCE before proceeding with analysis  of BAM files.\n\nUSAGE (default):\n\nfastaToCpg(FASTAfilename)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.  FastaToCpG(FASTAfilename,'maxChrNum',23)\n\nMADATORY INPUT:\n\nFASTAfilename                Full path of FASTA-formatted reference genome to which                available BAM files have been aligned to.\n\nOPTIONAL INPUTS:\n\noutdir                Path where the output will be stored at.          Default: \"REFGENEDIR\"\n\nmaxChrNum                Maximum number of chromosomes to be processed. The                function will process chromosomes 1,2,...,maxChrNum.                Default value: 22\n\nwsize                Window size used in CpG density calculations.                Default value: 1000\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"convertBAMtoBits(bamFilenames,phenoName;\n                          reference_path=\"./genome/\",\n                          bamfile_path=\"./indexedBAMfiles/\",\n                          matrices_path=\"./matrices/\",\n                          estimation_path=\"./estimation/\",\n                          outdir=\"./output/\",\n                          pairedEnds=true,\n                          numBasesToTrim=0,\n                          minCpGsReqToModel=10,\n                          regionSize=3000,\n                          boundaryConditions=false,\n                          MSIflag=false,\n                          ESIflag=false,\n                          MCflag=false,\n                          subRegionSize=150,\n                          chr_nums=1:22,\n                          numProcessors=nprocs())","category":"page"},{"location":"#InformMe.convertBAMtoBits-Tuple{Any,Any}","page":"Home","title":"InformMe.convertBAMtoBits","text":"Runs the entire MethToBits pipeline.\n\nThis function depends on a working instalation of SAMtools that is on  the system path $PATH.\n\nBefore running this function, FastaToCpG.m must be run ONCE.\n\nUSAGE (default):\n\nconvertBAMtoBits(bamFilenames,phenoName)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.\n\nmatrixFromBam(bamprefix,chrnum,'reference_path','/path/to/ref')\n\nMADATORY INPUTS:\n\nbamFilenames                 A comma seperated string with list of input bam file                 names without the \".bam\" extension. These                 files must be sorted from the least to the greatest base                 pair position along the reference sequence and must be                 indexed (i.e., the associated BAI file must be available).                 The file name must not contain \".\" characters, but can                 contain \"_\" instead. Moreover, the file name should be                 unique.\n\nphenoName                 A string which will be the unique identifier of this                  sample/model that is built.\n\nOPTIONAL INPUTS:\n\nreference_path                 Path to the root subdirectory where the outputs of this                 function are stored.                 Default value: \"./genome/\"\n\nbamfile_path                 Path to the subdirectory where the BAM file is located.                 Default value: \"./indexedBAMfiles/\"\n\nmatrices_path                 Path to the subdirectory where the output of this function                 is stored.                 Default value: \"./matrices/\"\n\nestimation_path                 A string that specifies the path to the directory that                 contains the results of parameter estimation performed                 by estParamsForChr.jl.                 Default value: \"./estimation/\"\n\noutdir                A string that specifies the path of the directory in which                the methylation analysis results are written.                Default value: \"./output/\"\n\npairedEnds                 Flag for paired end read support. A value of 1 indicates                 that the sequencer employed paired end reads, whereas a                 value of 0 indicates that the sequencer employed single                 end reads.                 Default value: true\n\nnumBasesToTrim                 A vector of integers specifying how many bases should be                 trimmed from the begining of each read. If the vector                 contains two integers, then the first integer specifies                 how many bases to trim from the first read in a read pair,                 whereas the second integer specifies how many bases should                 be trimmed from the second read in the pair. If the                 vector contains one integer, then all reads will have                 that number of bases trimmed from the beginning of the                 read. If no bases are to be trimmed, then this input                 must be set to 0.                 Default value: 0\n\nregionSize                 The size of the genomic regions for which methylation                 information is produced (in number of base pairs).                 Default value: 3000\n\nminCpGsReqToModel                 The minimum number of CpG sites within a genomic region                 required to perform statistical estimation.                 Default value: 10\n\nboundaryConditions                 Flag to decide if boundary conditions should be estimated                 freely in MLE.                 Default value: false\n\nMSIflag                Flag that determines whether this function performs                computation of the methylation sensitivity index (MSI).                false: no MSI computation.                true: allow MSI computation.                Default value: false\n\nESIflag                Flag that determines whether this function performs                computation of the entropic sensitivity index (ESI).                false: no ESI computation.                true: allow ESI computation.                Default value: false\n\nMCflag                Flag that determines whether this function performs                computation of turnover ratios, CpG entropies, capacities,                and relative dissipated energies of methylation                channels (MCs).                false: no MC computations.                true: allow MC computations.                Default value: false\n\nchr_nums               A vector with the chromosomes to be processed (without                \"chr\" string).                Default value: 1:22\n\nnumProcessors               The number of processors to use in the computations.               Note that julia must be started as \"julia -p 4\" if               four processors are desired. The nprocs() function               tells how many cores are available in julia, and               we default to use them all.               Default value: nprocs() \n\nThe default values of regionSize and minCpGsReqToModel should only be  changed by an expert with a detailed understanding of the code and the  methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"matrixFromBam(bamFilename,chr_num;\n                       reference_path=\"./genome/\",\n                       bamfile_path=\"./indexedBAMfiles/\",\n                       matrices_path=\"./matrices/\",\n                       pairedEnds=true,\n                       numBasesToTrim=0,\n                       regionSize=3000,\n                       minCpGsReqToModel=10)","category":"page"},{"location":"#InformMe.matrixFromBam-Tuple{Any,Any}","page":"Home","title":"InformMe.matrixFromBam","text":"This function processes a BAM file with aligned reads to a reference  genome and produces methylation information for nonoverlapping genomic  regions (containing the same number of base pairs) in a given chromosome.  The final output for each genomic region is a matrix with -1,0,1 values.  Each row of the matrix is a methylation read, whereas each column  represents a CpG site within the genomic region. A value of -1 indicates  no methylation information is available for the CPG site, 0 indicates  that the CpG site is unmethylated, and 1 indicates that the CpG site  is methylated.\n\nThis function depends on a working instalation of SAMtools that is on  the system path PATH.\n\nBefore running this function, FastaToCpG.m must be run ONCE.\n\nUSAGE (default):\n\nmatrixFromBam(bamprefix,chrnum)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.  matrixFromBam(bamprefix,chrnum; reference_path=\"/path/to/ref\")\n\nMADATORY INPUTS:\n\nbamprefix                 Prefix of the BAM file (without the .bam extension). This                 file must be sorted from the least to the greatest base                 pair position along the reference sequence and must be                 indexed (i.e., the associated BAI file must be available).                 The file name must not contain \".\" characters, but can                 contain \"\" instead. Moreover, the file name should be                 unique.\n\nchr_num                 Number representing the chromosome to be processed.\n\nOPTIONAL INPUTS:\n\nreference_path                 Path to the root subdirectory where the outputs of this                 function are stored.                 Default value: \"./genome/\"\n\nbamfile_path                 Path to the subdirectory where the BAM file is located.                 Default value: \"./indexedBAMfiles/\"\n\nmatrices_path                 Path to the subdirectory where the output of this function                 is stored.                 Default value: \"./matrices/\"\n\npairedEnds                 Flag for paired end read support. A value of true indicates                 that the sequencer employed paired end reads, whereas a                 value of false indicates that the sequencer employed single                 end reads.                 Default value: true\n\nnumBasesToTrim                 A vector of integers specifying how many bases should be                 trimmed from the begining of each read. If the vector                 contains two integers, then the first integer specifies                 how many bases to trim from the first read in a read pair,                 whereas the second integer specifies how many bases should                 be trimmed from the second read in the pair. If the                 vector contains one integer, then all reads will have                 that number of bases trimmed from the beginning of the                 read. If no bases are to be trimmed, then this input                 must be set to 0.                 Default value: 0\n\nregionSize                 The size of the genomic regions for which methylation                 information is produced (in number of base pairs).                 Default value: 3000\n\nminCpGsReqToModel                 The minimum number of CpG sites within a genomic region                 required to perform statistical estimation.                 Default value: 10\n\nThe default values of regionSize and minCpGsReqToModel should only be  changed by an expert with a detailed understanding of the code and the  methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"estParamsForChr(mat_files,phenoName,matrices_path,reference_path,chr_num;\n                         estimation_path=\"./results/\",\n                         regionSize=3000,\n                         boundaryConditions=false)","category":"page"},{"location":"#InformMe.estParamsForChr-NTuple{5,Any}","page":"Home","title":"InformMe.estParamsForChr","text":"This function takes a list of BAM files (which correspond to the same  phenotype) and performs statistical model estimation within a specific  chromosome of interest. The function can be used on a computing cluster  to break the work of model estimation to many independent parallel job  processes. This is performed only after matrixFromBam.m.\n\nUSAGE (default):\n\nestParamsForChr(matfiles,prefix,matricespath,referencepath,chrnum)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.  estParamsForChr(matfiles,prefix,matricespath,referencepath,chrnum,                  regionSize=2000)\n\nMANDATORY INPUTS:\n\nmat_files         All the .mat files to be included in the model. This can be a         single .mat file or multiple files in the form of a comma-sepa-         rated list of files.\n\nprefix         A string that specifies the name of the modeled phenotype.         The output files produced will contain this prefix.\n\nmatrices_path         A string that specifies the path to the directory that         where the output will be stored.\n\nreference_path         A string that specifies the path to the directory that         contains the results of analysis of the reference genome         performed by FastaToCpG.m as well as the results of         methylation calling performed by matrixFromBam.m.\n\nchr_num         Chromosome number 1 to 22 (in humans) specifying the         chromosome for which statistical estimation must be         performed.\n\nOPTIONAL INPUTS:\n\nregionSize         The size of the genomic region used for parameter         estimation (in number of base pairs).         Default value: 3000\n\nboundaryConditions         Flag to decide if boundary conditions should be estimated         freely in MLE.         Default value: false\n\nThe default value of regionSize should only be changed by an expert with  a detailed understanding of the code and the methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":" methAnalysisForChr(phenoName,chr_num,reference_path,estimation_path;\n                        outdir=\"./results/\",\n                        MSIflag=false,\n                        ESIflag=false,\n                        MCflag=false,\n                        regionSize=3000,\n                        subRegionSize=150)","category":"page"},{"location":"#InformMe.methAnalysisForChr-NTuple{4,Any}","page":"Home","title":"InformMe.methAnalysisForChr","text":"This function performs methylation analysis of a given chromosome in a  single phenotype. The function can be used on a computing cluster to  break the analysis work to many independent parallel job processes.  This is performed only after estParamsForChr.m in the Modeling  subdirectory is run to build the Ising models for the phenotype.\n\nUSAGE (default):\n\nmethAnalysisForChr(prefix,chrnum,referencepath,estimation_path)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.  methAnalysisForChr(prefix,chrnum,referencepath,estimation_path,                     outdir=\"/path/to/output\")\n\nMANDATORY INPUTS:\n\nprefix                A string that specifies the name of the phenotype to be                analyzed.\n\nchr_num                Chromosome number (1 to 22 in humans) specifying the                chromosome for which methylation analyis must be                performed.\n\nreference_path                A string that specifies the path to the directory that                contains the results of analysis of the reference genome                performed by FastaToCpG.m as well as the results of                methylation calling performed by matrixFromBam.jl.\n\nestimation_path                A string that specifies the path to the directory that                contains the results of parameter estimation performed                by estParamsForChr.jl.\n\nOPTIONAL INPUTS:\n\noutdir                A string that specifies the path of the directory in which                the methylation analysis results are written.                Default value: \"./results/\"\n\nMSIflag                Flag that determines whether this function performs                computation of the methylation sensitivity index (MSI).                false: no MSI computation.                true: allow MSI computation.                Default value: false\n\nESIflag                Flag that determines whether this function performs                computation of the entropic sensitivity index (ESI).                false: no ESI computation.                true: allow ESI computation.                Default value: false\n\nMCflag                Flag that determines whether this function performs                computation of turnover ratios, CpG entropies, capacities,                and relative dissipated energies of methylation                channels (MCs).                false: no MC computations.                true: allow MC computations.                Default value: false\n\nregionSize                The size of the genomic regions used for parameter                estimation (in number of base pairs).                Default value: 3000\n\nsubregionSize                The size of the subregions of a genomic region used                for methylation analysis (in number of base pairs).                The ratio regionSize/subregionSize must be an integer.                Default value: 150\n\nThe default values of regionSize and subregionSize should only be  changed by an expert with a detailed understanding of the code and  the methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"makeBedsForMethAnalysis(phenoName,analysis_path,reference_path;\n                              outdir=\"./singleMethAnalysisToBed_out/\",\n                              chrs=[string(\"chr\",i) for i=1:22],\n                              #minChrNum=1,\n                              #maxChrNum=22,\n                              MSIflag=false,\n                              ESIflag=false,\n                              MCflag=false,\n                              thresh=0.4,\n                              regionSize=3000,\n                              subregionSize=150)","category":"page"},{"location":"#InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}","page":"Home","title":"InformMe.makeBedsForMethAnalysis","text":"This function makes BED files for the methylation analysis results  obtained by means of MethAnalysisForChr.m for a single phenotype.\n\nUSAGE (default):\n\nmakeBedsForMethAnalysis(prefix,analysispath,referencepath)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.  makeBedsForMethAnalysis(prefix,analysispath,referencepath,'outdir',     '/path/to/output')\n\nMANDATORY INPUTS:\n\nprefix                A string that specifies the name of the phenotype.\n\nanalysis_path                A string that specifies the path of the directory in which                the model was constructed.         Default: \"$INTERMEDIATE\"\n\nreference_path                A string that specifies the path to the directory that                contains the results of analysis of the reference genome                performed by FastaToCpG.m as well as the results of                methylation calling performed by matrixFromBam.jl.         Default: \"$REFGENEDIR\"\n\nOPTIONAL INPUTS:\n\nminChrNum                A number specifying the starting chromosome that will be                included in the BED files.                Default value: 1\n\nmaxChrNum                A number specifying the last chromosome that will be                included in the output BED files. Must be                maxChrNum >= minChrNum.                Default value: 22\n\noutdir                A string that specifies the path of the directory in which                the output BED files are written.                Default value './'\n\nMSIflag                Flag that determines whether this function performs                computation of the methylation sensitivity index (MSI).                false: no MSI computation.                true: allow MSI computation.                Default value: false\n\nESIflag                Flag that determines whether this function performs                computation of the entropic sensitivity index (ESI).                false: no ESI computation.                true: allow ESI computation.                Default value: false\n\nMCflag                Flag that determines whether this function performs                computation of turnover ratios, CpG entropies, capacities,                and relative dissipated energies of methylation                channels (MCs).                false: no MC computations.                true: allow MC computations.                Default value: false\n\nthresh                A scalar used as a threshold in methylation-based                classification.                Default value: 0.4\n\nregionSize                The size of the genomic regions used for parameter                estimation (in number of base pairs).                Default value: 3000\n\nsubregionSize                The size of the subregions of a genomic region used                for methylation analysis (in number of base pairs).                The ratio regionSize/subregionSize must be an integer.                Default value: 150\n\nThe default values of thresh, regionSize, and subregionSize should only  be changed by an expert with a detailed understanding of the code and  the methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"diffMethAnalysisToBed(phenoName_1,phenoName_2,analysis_path_1,analysis_path_2,reference_path,\n                                  outdir=\"../makeBedsForDiffMethAnalysis_out/\",\n                                  chrs=[string(\"chr\",i) for i=1:22],\n                                  #  minChrNum=1,\n                                  #  maxChrNum=22,\n                                  MSIflag=false,\n                                  ESIflag=false,\n                                  MCflag=false,\n                                  regionSize=3000,\n                                  subregionSize=150,\n                                  minNumCpG=2,\n                                  thresh=0.55,\n                                  threshDMU=[-1,-0.55,-0.1,0.1,0.55,1],\n                                  threshDEU=[-1,-.5,-.3,-.05,.05,.3,.5,1])","category":"page"},{"location":"#InformMe.diffMethAnalysisToBed","page":"Home","title":"InformMe.diffMethAnalysisToBed","text":"This function makes BED files for the differential version of the  methylation analysis results obtained by means of MethAnalysisForChr.m  applied on two dinstict phenotypes.\n\nUSAGE (default):\n\nmakeBedsForDiffMethAnalysis(prefix1,prefix2,analysispath1,...     analysispath2,reference_path)\n\nMANDATORY INPUTS:\n\nprefix_X                Strings with the first string specifying                the name of the first phenotype and the second string                specifying the name of the second phenotype used for                differential methylation analysis. Both phenotypes                must have already been analyzed with methAnalysisForChr.m.\n\nanalysispathX                A string that specifies the path of the directory in which                the methylation analysis results obtained by                MethAnalysisForChr.m are stored.         Default: \"$INTERMEDIATE\"\n\nreference_path                A string that specifies the path to the directory that                contains the results of analysis of the reference genome                performed by FastaToCpG.m as well as the results of                methylation calling performed by matrixFromBam.m.         Default: \"$REFGENEDIR\"\n\nOPTIONAL INPUTS:\n\nminChrNum                A number specifying the starting chromosome that will be                included in the BED files.                Default value: 1\n\nmaxChrNum                A number specifying the last chromosome that will be                included in the outut BED files. Must be                maxChrNum >= minChrNum.                Default value: 22\n\noutdir                A string that specifies the path of the directory in which                the output BED files are written.                Default value \"$INTERMEDIATE\"\n\nMSIflag                Flag that determines whether this function performs                computation of the methylation sensitivity index (MSI).                0: no MSI computation.                1: allow MSI computation.                Default value: 0\n\nESIflag                Flag that determines whether this function performs                computation of the entropic sensitivity index (ESI).                0: no ESI computation.                1: allow ESI computation.                Default value: 0\n\nMCflag                Flag that determines whether this function performs                computation of turnover ratios, CpG entropies, capacities,                and relative dissipated energies of methylation                channels (MCs).                0: no MC computations.                1: allow MC computations.                Default value: 0\n\nregionSize                The size of the genomic regions used for parameter                estimation (in number of base pairs).                Default value: 3000\n\nsubregionSize                The size of the subregions of a genomic region used                for methylation analysis (in number of base pairs).                The ratio regionSize/subregionSize must be an integer.                Default value: 150\n\nminNumCpG     The minimum number of CpG sites within an analysis                subregion required for performing full methylation-based                differential classification.                Default value: 2\n\nthresh\n\n           A scalar used as a threshold in methylation-based\n           differential classification.\n           Default value: 0.55\n\nthreshDMU                A 1x6 vector containing threshold values used for                methylation-based differential classification.                Default value: [-1,-0.55,-0.1,0.1,0.55,1]\n\nthreshDEU                A 1x8 vector containing threshold values used for                entropy-based differential classification.                Default value: [-1,-0.5,-0.3,-0.05,0.05,0.3,0.5,1]\n\nThe default values of regionSize, subregionSize, minNumCpG, thresh,  threshDMU, and threshDEU should only be changed by an expert with  a detailed understanding of the code and the methods used.\n\n\n\n\n\n","category":"function"},{"location":"#Index-1","page":"Home","title":"Index","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"","category":"page"}]
}
