<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · InformMe.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">InformMe.jl</span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Functions-1"><span>Functions</span></a></li><li><a class="tocitem" href="#Index-1"><span>Index</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/master/docs/src/index.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="InformMe.jl-Documentation-1"><a class="docs-heading-anchor" href="#InformMe.jl-Documentation-1">InformMe.jl Documentation</a><a class="docs-heading-anchor-permalink" href="#InformMe.jl-Documentation-1" title="Permalink"></a></h1><ul><li><a href="#InformMe.jl-Documentation-1">InformMe.jl Documentation</a></li><ul><li><a href="#Functions-1">Functions</a></li><li><a href="#Index-1">Index</a></li></ul></ul><h2 id="Functions-1"><a class="docs-heading-anchor" href="#Functions-1">Functions</a><a class="docs-heading-anchor-permalink" href="#Functions-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="InformMe.fastaToCpG-Tuple{Any}" href="#InformMe.fastaToCpG-Tuple{Any}"><code>InformMe.fastaToCpG</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function is used to analyze a reference genome in order to find  and store the locations of all CpG sites within each chromosome and  to compute the CpG densities at each CpG site as well as the distances  between neighboring CpG sites. A 1-based coordinate system is used,  in which the first base is assigned to position 1 and the location  of a CpG site is defined by the position of the C nucleotide on the  forward strand of the reference genome.</p><p>This function must be run ONLY ONCE before proceeding with analysis  of BAM files.</p><p><strong>USAGE (default):</strong></p><p><code>fastaToCpg(FASTAfilename)</code></p><p><strong>USAGE (optional):</strong></p><p>Example of optional usage with additional input parameters.</p><p><code>FastaToCpG(FASTAfilename,outdir=&quot;/path/to/outputdir/&quot;)</code></p><p><strong>MADATORY INPUT:</strong></p><p><code>FASTAfilename</code></p><pre><code class="language-none">           Full path of FASTA-formatted reference genome to which
           available BAM files have been aligned to.</code></pre><p><strong>OPTIONAL INPUTS:</strong></p><p><code>outdir</code></p><pre><code class="language-none">           Path where the output will be stored at.
           Default value: &quot;./&quot;</code></pre><p><code>wsize</code></p><pre><code class="language-none">           Window size used in CpG density calculations.
           Default value: 1000</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/6f6bb81a5d0b193867f2474e5869e676af689da7/src/fastaToCpG.jl#LL20-L61">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.convertBAMtoBits-Tuple{Any,Any}" href="#InformMe.convertBAMtoBits-Tuple{Any,Any}"><code>InformMe.convertBAMtoBits</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Runs the entire MethToBits pipeline.</p><p>This function depends on a working instalation of SAMtools that is on  the system path PATH.</p><p>Before running this function, FastaToCpG.m must be run ONCE.</p><p><strong>USAGE (default):</strong></p><p><code>convertBAMtoBits(bamFilenames,phenoName)</code></p><p><strong>USAGE (optional):</strong></p><p>Example of optional usage with additional input parameters.</p><p><code>matrixFromBam(bam_prefix,chr_num,reference_path=&quot;/path/to/ref&quot;)</code></p><p><strong>MADATORY INPUTS:</strong></p><p><code>bamFilenames</code></p><pre><code class="language-none">            A comma seperated string with list of input bam file
            names without the &quot;.bam&quot; extension. These
            files must be sorted from the least to the greatest base
            pair position along the reference sequence and must be
            indexed (i.e., the associated BAI file must be available).
            The file name must not contain &quot;.&quot; characters, but can
            contain &quot;_&quot; instead. Moreover, the file name should be
            unique.</code></pre><p><code>phenoName</code></p><pre><code class="language-none">            A string which will be the unique identifier of this 
            sample/model that is built.</code></pre><p>##OPTIONAL INPUTS:</p><p><code>reference_path</code></p><pre><code class="language-none">            Path to the root subdirectory where the outputs of this
            function are stored.
            Default value: &quot;./genome/&quot;</code></pre><p><code>bamfile_path</code></p><pre><code class="language-none">            Path to the subdirectory where the BAM file is located.
            Default value: &quot;./indexedBAMfiles/&quot;</code></pre><p><code>matrices_path</code></p><pre><code class="language-none">            Path to the subdirectory where the output of this function
            is stored.
            Default value: &quot;./matrices/&quot;</code></pre><p><code>estimation_path</code></p><pre><code class="language-none">            A string that specifies the path to the directory that
            contains the results of parameter estimation performed
            by estParamsForChr.jl.
            Default value: &quot;./estimation/&quot;</code></pre><p><code>outdir</code></p><pre><code class="language-none">           A string that specifies the path of the directory in which
           the methylation analysis results are written.
           Default value: &quot;./output/&quot;</code></pre><p><code>pairedEnds</code></p><pre><code class="language-none">            Flag for paired end read support. A value of 1 indicates
            that the sequencer employed paired end reads, whereas a
            value of 0 indicates that the sequencer employed single
            end reads.
            Default value: true</code></pre><p><code>numBasesToTrim</code></p><pre><code class="language-none">            A vector of integers specifying how many bases should be
            trimmed from the begining of each read. If the vector
            contains two integers, then the first integer specifies
            how many bases to trim from the first read in a read pair,
            whereas the second integer specifies how many bases should
            be trimmed from the second read in the pair. If the
            vector contains one integer, then all reads will have
            that number of bases trimmed from the beginning of the
            read. If no bases are to be trimmed, then this input
            must be set to 0.
            Default value: 0</code></pre><p><code>regionSize</code></p><pre><code class="language-none">            The size of the genomic regions for which methylation
            information is produced (in number of base pairs).
            Default value: 3000</code></pre><p><code>minCpGsReqToModel</code></p><pre><code class="language-none">            The minimum number of CpG sites within a genomic region
            required to perform statistical estimation.
            Default value: 10</code></pre><p><code>boundaryConditions</code></p><pre><code class="language-none">            Flag to decide if boundary conditions should be estimated
            freely in MLE.
            Default value: false</code></pre><p><code>MSIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the methylation sensitivity index (MSI).
           false: no MSI computation.
           true: allow MSI computation.
           Default value: false</code></pre><p><code>ESIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the entropic sensitivity index (ESI).
           false: no ESI computation.
           true: allow ESI computation.
           Default value: false</code></pre><p><code>MCflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of turnover ratios, CpG entropies, capacities,
           and relative dissipated energies of methylation
           channels (MCs).
           false: no MC computations.
           true: allow MC computations.
           Default value: false</code></pre><p><code>chr_nums</code></p><pre><code class="language-none">          A vector with the chromosomes to be processed (without 
          &quot;chr&quot; string). 
          Default value: 1:22</code></pre><p><code>numProcessors</code></p><pre><code class="language-none">          The number of processors to use in the computations.
          Note that julia must be started as &quot;julia -p 4&quot; if
          four processors are desired. The nprocs() function
          tells how many cores are available in julia, and
          we default to use them all.
          Default value: nprocs()</code></pre><p>The default values of regionSize and minCpGsReqToModel should only be  changed by an expert with a detailed understanding of the code and the  methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/6f6bb81a5d0b193867f2474e5869e676af689da7/src/convertBAMtoBits.jl#LL20-L175">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.matrixFromBam-Tuple{Any,Any}" href="#InformMe.matrixFromBam-Tuple{Any,Any}"><code>InformMe.matrixFromBam</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function processes a BAM file with aligned reads to a reference  genome and produces methylation information for nonoverlapping genomic  regions (containing the same number of base pairs) in a given chromosome.  The final output for each genomic region is a matrix with -1,0,1 values.  Each row of the matrix is a methylation read, whereas each column  represents a CpG site within the genomic region. A value of -1 indicates  no methylation information is available for the CPG site, 0 indicates  that the CpG site is unmethylated, and 1 indicates that the CpG site  is methylated.</p><p>This function depends on a working instalation of SAMtools that is on  the system path PATH.</p><p>Before running this function, FastaToCpG.m must be run ONCE.</p><p><strong>USAGE (default):</strong></p><p><code>matrixFromBam(bam_prefix,chr_num)</code></p><p><strong>USAGE (optional):</strong></p><p>Example of optional usage with additional input parameters.</p><p><code>matrixFromBam(bam_prefix,chr_num; reference_path=&quot;/path/to/ref&quot;)</code></p><p><strong>MADATORY INPUTS:</strong></p><p><code>bam_prefix</code></p><pre><code class="language-none">            Prefix of the BAM file (without the .bam extension). This
            file must be sorted from the least to the greatest base
            pair position along the reference sequence and must be
            indexed (i.e., the associated BAI file must be available).
            The file name must not contain &quot;.&quot; characters, but can
            contain &quot;_&quot; instead. Moreover, the file name should be
            unique.</code></pre><p><code>chr_num</code></p><pre><code class="language-none">            Number representing the chromosome to be processed.</code></pre><p><strong>OPTIONAL INPUTS:</strong></p><p><code>reference_path</code></p><pre><code class="language-none">            Path to the root subdirectory where the outputs of this
            function are stored.
            Default value: &quot;./genome/&quot;</code></pre><p><code>bamfile_path</code></p><pre><code class="language-none">            Path to the subdirectory where the BAM file is located.
            Default value: &quot;./indexedBAMfiles/&quot;</code></pre><p><code>matrices_path</code></p><pre><code class="language-none">            Path to the subdirectory where the output of this function
            is stored.
            Default value: &quot;./matrices/&quot;</code></pre><p><code>pairedEnds</code></p><pre><code class="language-none">            Flag for paired end read support. A value of true indicates
            that the sequencer employed paired end reads, whereas a
            value of false indicates that the sequencer employed single
            end reads.
            Default value: true</code></pre><p><code>numBasesToTrim</code></p><pre><code class="language-none">            A vector of integers specifying how many bases should be
            trimmed from the begining of each read. If the vector
            contains two integers, then the first integer specifies
            how many bases to trim from the first read in a read pair,
            whereas the second integer specifies how many bases should
            be trimmed from the second read in the pair. If the
            vector contains one integer, then all reads will have
            that number of bases trimmed from the beginning of the
            read. If no bases are to be trimmed, then this input
            must be set to 0.
            Default value: 0</code></pre><p><code>regionSize</code></p><pre><code class="language-none">            The size of the genomic regions for which methylation
            information is produced (in number of base pairs).
            Default value: 3000</code></pre><p><code>minCpGsReqToModel</code></p><pre><code class="language-none">            The minimum number of CpG sites within a genomic region
            required to perform statistical estimation.
            Default value: 10</code></pre><p>The default values of regionSize and minCpGsReqToModel should only be  changed by an expert with a detailed understanding of the code and the  methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/6f6bb81a5d0b193867f2474e5869e676af689da7/src/matrixFromBam.jl#LL21-L119">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.estParamsForChr-NTuple{5,Any}" href="#InformMe.estParamsForChr-NTuple{5,Any}"><code>InformMe.estParamsForChr</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function takes a list of BAM files (which correspond to the same  phenotype) and performs statistical model estimation within a specific  chromosome of interest. The function can be used on a computing cluster  to break the work of model estimation to many independent parallel job  processes. This is performed only after matrixFromBam.m.</p><p><strong>USAGE (default):</strong></p><p><code>estParamsForChr(mat_files,prefix,matrices_path,reference_path,chr_num)</code></p><p><strong>USAGE (optional):</strong></p><p>Example of optional usage with additional input parameters.</p><p><code>estParamsForChr(mat_files,prefix,matrices_path,reference_path,chr_num,                  regionSize=2000)</code></p><p>MANDATORY INPUTS:</p><p><code>mat_files</code></p><pre><code class="language-none">    All the .mat files to be included in the model. This can be a
    single .mat file or multiple files in the form of a comma-sepa-
    rated list of files.</code></pre><p><code>prefix</code></p><pre><code class="language-none">    A string that specifies the name of the modeled phenotype.
    The output files produced will contain this prefix.</code></pre><p><code>matrices_path</code></p><pre><code class="language-none">    A string that specifies the path to the directory that
    where the output will be stored.</code></pre><p><code>reference_path</code></p><pre><code class="language-none">    A string that specifies the path to the directory that
    contains the results of analysis of the reference genome
    performed by FastaToCpG.m as well as the results of
    methylation calling performed by matrixFromBam.m.</code></pre><p><code>chr_num</code></p><pre><code class="language-none">    Chromosome number 1 to 22 (in humans) specifying the
    chromosome for which statistical estimation must be
    performed.</code></pre><p><strong>OPTIONAL INPUTS:</strong></p><p><code>regionSize</code></p><pre><code class="language-none">    The size of the genomic region used for parameter
    estimation (in number of base pairs).
    Default value: 3000</code></pre><p><code>boundaryConditions</code></p><pre><code class="language-none">    Flag to decide if boundary conditions should be estimated
    freely in MLE.
    Default value: false</code></pre><p>The default value of regionSize should only be changed by an expert with  a detailed understanding of the code and the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/6f6bb81a5d0b193867f2474e5869e676af689da7/src/estParamsForChr.jl#LL21-L86">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.methAnalysisForChr-NTuple{4,Any}" href="#InformMe.methAnalysisForChr-NTuple{4,Any}"><code>InformMe.methAnalysisForChr</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function performs methylation analysis of a given chromosome in a  single phenotype. The function can be used on a computing cluster to  break the analysis work to many independent parallel job processes.  This is performed only after estParamsForChr.m in the Modeling  subdirectory is run to build the Ising models for the phenotype.</p><p><strong>USAGE (default):</strong></p><p><code>methAnalysisForChr(prefix,chr_num,reference_path,estimation_path)</code></p><p><strong>USAGE (optional):</strong></p><p>Example of optional usage with additional input parameters.</p><p><code>methAnalysisForChr(prefix,chr_num,reference_path,estimation_path,                     outdir=&quot;/path/to/output&quot;)</code></p><p><strong>MANDATORY INPUTS:</strong></p><p><code>prefix</code></p><pre><code class="language-none">           A string that specifies the name of the phenotype to be
           analyzed.</code></pre><p><code>chr_num</code></p><pre><code class="language-none">           Chromosome number (1 to 22 in humans) specifying the
           chromosome for which methylation analyis must be
           performed.</code></pre><p><code>reference_path</code></p><pre><code class="language-none">           A string that specifies the path to the directory that
           contains the results of analysis of the reference genome
           performed by FastaToCpG.m as well as the results of
           methylation calling performed by matrixFromBam.jl.</code></pre><p><code>estimation_path</code></p><pre><code class="language-none">           A string that specifies the path to the directory that
           contains the results of parameter estimation performed
           by estParamsForChr.jl.</code></pre><p><strong>OPTIONAL INPUTS:</strong></p><p><code>outdir</code></p><pre><code class="language-none">           A string that specifies the path of the directory in which
           the methylation analysis results are written.
           Default value: &quot;./results/&quot;</code></pre><p><code>MSIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the methylation sensitivity index (MSI).
           false: no MSI computation.
           true: allow MSI computation.
           Default value: false</code></pre><p><code>ESIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the entropic sensitivity index (ESI).
           false: no ESI computation.
           true: allow ESI computation.
           Default value: false</code></pre><p><code>MCflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of turnover ratios, CpG entropies, capacities,
           and relative dissipated energies of methylation
           channels (MCs).
           false: no MC computations.
           true: allow MC computations.
           Default value: false</code></pre><p><code>regionSize</code></p><pre><code class="language-none">           The size of the genomic regions used for parameter
           estimation (in number of base pairs).
           Default value: 3000</code></pre><p><code>subregionSize</code></p><pre><code class="language-none">           The size of the subregions of a genomic region used
           for methylation analysis (in number of base pairs).
           The ratio regionSize/subregionSize must be an integer.
           Default value: 150</code></pre><p>The default values of regionSize and subregionSize should only be  changed by an expert with a detailed understanding of the code and  the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/6f6bb81a5d0b193867f2474e5869e676af689da7/src/methAnalysisForChr.jl#LL21-L116">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}" href="#InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}"><code>InformMe.makeBedsForMethAnalysis</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function makes BED files for the methylation analysis results  obtained by means of MethAnalysisForChr.m for a single phenotype.</p><p><strong>USAGE (default):</strong></p><p><code>makeBedsForMethAnalysis(prefix,analysis_path,reference_path)</code></p><p><strong>USAGE (optional):</strong></p><p>Example of optional usage with additional input parameters.</p><p><code>makeBedsForMethAnalysis(prefix,analysis_path,reference_path,                          outdir=&quot;/path/to/output&quot;)</code></p><p><strong>MANDATORY INPUTS:</strong></p><p><code>prefix</code></p><pre><code class="language-none">           A string that specifies the name of the phenotype.</code></pre><p><code>analysis_path</code></p><pre><code class="language-none">           A string that specifies the path of the directory in which
           the model was constructed.</code></pre><p><code>reference_path</code></p><pre><code class="language-none">           A string that specifies the path to the directory that
           contains the results of analysis of the reference genome
           performed by FastaToCpG.m as well as the results of
           methylation calling performed by matrixFromBam.jl.</code></pre><p><strong>OPTIONAL INPUTS:</strong></p><p><code>chrs</code></p><pre><code class="language-none">          A vector of strings for the chromosomes to output to the
          final bed files. Default value: `[string(&quot;chr&quot;,i) for i=1:22]`</code></pre><p><code>outdir</code></p><pre><code class="language-none">           A string that specifies the path of the directory in which
           the output BED files are written.
           Default value: &quot;./&quot;</code></pre><p><code>MSIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the methylation sensitivity index (MSI).
           false: no MSI computation.
           true: allow MSI computation.
           Default value: false</code></pre><p><code>ESIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the entropic sensitivity index (ESI).
           false: no ESI computation.
           true: allow ESI computation.
           Default value: false</code></pre><p><code>MCflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of turnover ratios, CpG entropies, capacities,
           and relative dissipated energies of methylation
           channels (MCs).
           false: no MC computations.
           true: allow MC computations.
           Default value: false</code></pre><p><code>thresh</code></p><pre><code class="language-none">           A scalar used as a threshold in methylation-based
           classification.
           Default value: 0.4</code></pre><p><code>regionSize</code></p><pre><code class="language-none">           The size of the genomic regions used for parameter
           estimation (in number of base pairs).
           Default value: 3000</code></pre><p><code>subregionSize</code></p><pre><code class="language-none">           The size of the subregions of a genomic region used
           for methylation analysis (in number of base pairs).
           The ratio regionSize/subregionSize must be an integer.
           Default value: 150</code></pre><p>The default values of thresh, regionSize, and subregionSize should only  be changed by an expert with a detailed understanding of the code and  the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/6f6bb81a5d0b193867f2474e5869e676af689da7/src/makeBedsForMethAnalysis.jl#LL20-L115">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="InformMe.diffMethAnalysisToBed" href="#InformMe.diffMethAnalysisToBed"><code>InformMe.diffMethAnalysisToBed</code></a> — <span class="docstring-category">Function</span></header><section><div><p>This function makes BED files for the differential version of the  methylation analysis results obtained by means of MethAnalysisForChr.m  applied on two dinstict phenotypes.</p><p><strong>USAGE (default):</strong></p><p><code>makeBedsForDiffMethAnalysis(prefix_1,prefix_2,analysis_path_1,     analysis_path_2,reference_path)</code></p><p><strong>MANDATORY INPUTS:</strong></p><p><code>prefix_X</code></p><pre><code class="language-none">           Strings with the first string specifying
           the name of the first phenotype and the second string
           specifying the name of the second phenotype used for
           differential methylation analysis. Both phenotypes
           must have already been analyzed with methAnalysisForChr.jl.</code></pre><p><code>analysis_path_X</code></p><pre><code class="language-none">           A string that specifies the path of the directory in which
           the methylation analysis results obtained by
           MethAnalysisForChr.jl are stored.</code></pre><p><code>reference_path</code></p><pre><code class="language-none">           A string that specifies the path to the directory that
           contains the results of analysis of the reference genome
           performed by FastaToCpG.m as well as the results of
           methylation calling performed by matrixFromBam.jl.</code></pre><p><strong>OPTIONAL INPUTS:</strong></p><p><code>chrs</code></p><pre><code class="language-none">           A vector of strings of chromosomes to be output to the final
           bed files. Default value: `[string(&quot;chr&quot;,i) for i=1:22]`</code></pre><p><code>outdir</code></p><pre><code class="language-none">           A string that specifies the path of the directory in which
           the output BED files are written.
           Default value: &quot;./makeBedsForDiffMethAnalysis_out/&quot;</code></pre><p><code>MSIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the methylation sensitivity index (MSI).
           false: no MSI computation.
           true: allow MSI computation.
           Default value: false</code></pre><p><code>ESIflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of the entropic sensitivity index (ESI).
           false: no ESI computation.
           true: allow ESI computation.
           Default value: false</code></pre><p><code>MCflag</code></p><pre><code class="language-none">           Flag that determines whether this function performs
           computation of turnover ratios, CpG entropies, capacities,
           and relative dissipated energies of methylation
           channels (MCs).
           false: no MC computations.
           true: allow MC computations.
           Default value: false</code></pre><p><code>regionSize</code></p><pre><code class="language-none">           The size of the genomic regions used for parameter
           estimation (in number of base pairs).
           Default value: 3000</code></pre><p><code>subregionSize</code></p><pre><code class="language-none">           The size of the subregions of a genomic region used
           for methylation analysis (in number of base pairs).
           The ratio regionSize/subregionSize must be an integer.
           Default value: 150</code></pre><p><code>minNumCpG</code></p><pre><code class="language-none">           The minimum number of CpG sites within an analysis
           subregion required for performing full methylation-based
           differential classification.
           Default value: 2</code></pre><p><code>thresh</code></p><pre><code class="language-none">           A scalar used as a threshold in methylation-based
           differential classification.
           Default value: 0.55</code></pre><p><code>threshDMU</code></p><pre><code class="language-none">           A 1x6 vector containing threshold values used for
           methylation-based differential classification.
           Default value: [-1,-0.55,-0.1,0.1,0.55,1]</code></pre><p><code>threshDEU</code></p><pre><code class="language-none">           A 1x8 vector containing threshold values used for
           entropy-based differential classification.
           Default value: [-1,-0.5,-0.3,-0.05,0.05,0.3,0.5,1]</code></pre><p>The default values of regionSize, subregionSize, minNumCpG, thresh,  threshDMU, and threshDEU should only be changed by an expert with  a detailed understanding of the code and the methods used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GarrettJenkinson/InformMe.jl/blob/6f6bb81a5d0b193867f2474e5869e676af689da7/src/diffMethAnalysisToBed.jl#LL20-L134">source</a></section></article><h2 id="Index-1"><a class="docs-heading-anchor" href="#Index-1">Index</a><a class="docs-heading-anchor-permalink" href="#Index-1" title="Permalink"></a></h2><ul><li><a href="#InformMe.convertBAMtoBits-Tuple{Any,Any}"><code>InformMe.convertBAMtoBits</code></a></li><li><a href="#InformMe.diffMethAnalysisToBed"><code>InformMe.diffMethAnalysisToBed</code></a></li><li><a href="#InformMe.estParamsForChr-NTuple{5,Any}"><code>InformMe.estParamsForChr</code></a></li><li><a href="#InformMe.fastaToCpG-Tuple{Any}"><code>InformMe.fastaToCpG</code></a></li><li><a href="#InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}"><code>InformMe.makeBedsForMethAnalysis</code></a></li><li><a href="#InformMe.matrixFromBam-Tuple{Any,Any}"><code>InformMe.matrixFromBam</code></a></li><li><a href="#InformMe.methAnalysisForChr-NTuple{4,Any}"><code>InformMe.methAnalysisForChr</code></a></li></ul></article></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 3 December 2019 04:39">Tuesday 3 December 2019</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
