var documenterSearchIndex = {"docs":
[{"location":"#InformMe.jl-Documentation-1","page":"Home","title":"InformMe.jl Documentation","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"","category":"page"},{"location":"#Functions-1","page":"Home","title":"Functions","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"fastaToCpG(FASTAfilename;\n                    outdir=\"./\",\n                    wsize=1000)","category":"page"},{"location":"#InformMe.fastaToCpG-Tuple{Any}","page":"Home","title":"InformMe.fastaToCpG","text":"This function is used to analyze a reference genome in order to find  and store the locations of all CpG sites within each chromosome and  to compute the CpG densities at each CpG site as well as the distances  between neighboring CpG sites. A 1-based coordinate system is used,  in which the first base is assigned to position 1 and the location  of a CpG site is defined by the position of the C nucleotide on the  forward strand of the reference genome.\n\nThis function must be run ONLY ONCE before proceeding with analysis  of BAM files.\n\nUSAGE (default):\n\nfastaToCpg(FASTAfilename)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.\n\nFastaToCpG(FASTAfilename,outdir=\"/path/to/outputdir/\")\n\nMADATORY INPUT:\n\nFASTAfilename\n\n           Full path of FASTA-formatted reference genome to which\n           available BAM files have been aligned to.\n\nOPTIONAL INPUTS:\n\noutdir\n\n           Path where the output will be stored at.\n           Default value: \"./\"\n\nwsize\n\n           Window size used in CpG density calculations.\n           Default value: 1000\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"convertBAMtoBits(bamFilenames,phenoName;\n                          reference_path=\"./genome/\",\n                          bamfile_path=\"./indexedBAMfiles/\",\n                          matrices_path=\"./matrices/\",\n                          estimation_path=\"./estimation/\",\n                          outdir=\"./output/\",\n                          pairedEnds=true,\n                          numBasesToTrim=0,\n                          minCpGsReqToModel=10,\n                          regionSize=3000,\n                          boundaryConditions=false,\n                          MSIflag=false,\n                          ESIflag=false,\n                          MCflag=false,\n                          subRegionSize=150,\n                          chr_nums=1:22,\n                          numProcessors=nprocs())","category":"page"},{"location":"#InformMe.convertBAMtoBits-Tuple{Any,Any}","page":"Home","title":"InformMe.convertBAMtoBits","text":"Runs the entire MethToBits pipeline.\n\nThis function depends on a working instalation of SAMtools that is on  the system path PATH.\n\nBefore running this function, FastaToCpG.m must be run ONCE.\n\nUSAGE (default):\n\nconvertBAMtoBits(bamFilenames,phenoName)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.\n\nmatrixFromBam(bam_prefix,chr_num,reference_path=\"/path/to/ref\")\n\nMADATORY INPUTS:\n\nbamFilenames\n\n            A comma seperated string with list of input bam file\n            names without the \".bam\" extension. These\n            files must be sorted from the least to the greatest base\n            pair position along the reference sequence and must be\n            indexed (i.e., the associated BAI file must be available).\n            The file name must not contain \".\" characters, but can\n            contain \"_\" instead. Moreover, the file name should be\n            unique.\n\nphenoName\n\n            A string which will be the unique identifier of this \n            sample/model that is built.\n\n##OPTIONAL INPUTS:\n\nreference_path\n\n            Path to the root subdirectory where the outputs of this\n            function are stored.\n            Default value: \"./genome/\"\n\nbamfile_path\n\n            Path to the subdirectory where the BAM file is located.\n            Default value: \"./indexedBAMfiles/\"\n\nmatrices_path\n\n            Path to the subdirectory where the output of this function\n            is stored.\n            Default value: \"./matrices/\"\n\nestimation_path\n\n            A string that specifies the path to the directory that\n            contains the results of parameter estimation performed\n            by estParamsForChr.jl.\n            Default value: \"./estimation/\"\n\noutdir\n\n           A string that specifies the path of the directory in which\n           the methylation analysis results are written.\n           Default value: \"./output/\"\n\npairedEnds\n\n            Flag for paired end read support. A value of 1 indicates\n            that the sequencer employed paired end reads, whereas a\n            value of 0 indicates that the sequencer employed single\n            end reads.\n            Default value: true\n\nnumBasesToTrim\n\n            A vector of integers specifying how many bases should be\n            trimmed from the begining of each read. If the vector\n            contains two integers, then the first integer specifies\n            how many bases to trim from the first read in a read pair,\n            whereas the second integer specifies how many bases should\n            be trimmed from the second read in the pair. If the\n            vector contains one integer, then all reads will have\n            that number of bases trimmed from the beginning of the\n            read. If no bases are to be trimmed, then this input\n            must be set to 0.\n            Default value: 0\n\nregionSize\n\n            The size of the genomic regions for which methylation\n            information is produced (in number of base pairs).\n            Default value: 3000\n\nminCpGsReqToModel\n\n            The minimum number of CpG sites within a genomic region\n            required to perform statistical estimation.\n            Default value: 10\n\nboundaryConditions\n\n            Flag to decide if boundary conditions should be estimated\n            freely in MLE.\n            Default value: false\n\nMSIflag\n\n           Flag that determines whether this function performs\n           computation of the methylation sensitivity index (MSI).\n           false: no MSI computation.\n           true: allow MSI computation.\n           Default value: false\n\nESIflag\n\n           Flag that determines whether this function performs\n           computation of the entropic sensitivity index (ESI).\n           false: no ESI computation.\n           true: allow ESI computation.\n           Default value: false\n\nMCflag\n\n           Flag that determines whether this function performs\n           computation of turnover ratios, CpG entropies, capacities,\n           and relative dissipated energies of methylation\n           channels (MCs).\n           false: no MC computations.\n           true: allow MC computations.\n           Default value: false\n\nchr_nums\n\n          A vector with the chromosomes to be processed (without \n          \"chr\" string). \n          Default value: 1:22\n\nnumProcessors\n\n          The number of processors to use in the computations.\n          Note that julia must be started as \"julia -p 4\" if\n          four processors are desired. The nprocs() function\n          tells how many cores are available in julia, and\n          we default to use them all.\n          Default value: nprocs()\n\nThe default values of regionSize and minCpGsReqToModel should only be  changed by an expert with a detailed understanding of the code and the  methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"matrixFromBam(bamFilename,chr_num;\n                       reference_path=\"./genome/\",\n                       bamfile_path=\"./indexedBAMfiles/\",\n                       matrices_path=\"./matrices/\",\n                       pairedEnds=true,\n                       numBasesToTrim=0,\n                       regionSize=3000,\n                       minCpGsReqToModel=10)","category":"page"},{"location":"#InformMe.matrixFromBam-Tuple{Any,Any}","page":"Home","title":"InformMe.matrixFromBam","text":"This function processes a BAM file with aligned reads to a reference  genome and produces methylation information for nonoverlapping genomic  regions (containing the same number of base pairs) in a given chromosome.  The final output for each genomic region is a matrix with -1,0,1 values.  Each row of the matrix is a methylation read, whereas each column  represents a CpG site within the genomic region. A value of -1 indicates  no methylation information is available for the CPG site, 0 indicates  that the CpG site is unmethylated, and 1 indicates that the CpG site  is methylated.\n\nThis function depends on a working instalation of SAMtools that is on  the system path PATH.\n\nBefore running this function, FastaToCpG.m must be run ONCE.\n\nUSAGE (default):\n\nmatrixFromBam(bam_prefix,chr_num)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.\n\nmatrixFromBam(bam_prefix,chr_num; reference_path=\"/path/to/ref\")\n\nMADATORY INPUTS:\n\nbam_prefix\n\n            Prefix of the BAM file (without the .bam extension). This\n            file must be sorted from the least to the greatest base\n            pair position along the reference sequence and must be\n            indexed (i.e., the associated BAI file must be available).\n            The file name must not contain \".\" characters, but can\n            contain \"_\" instead. Moreover, the file name should be\n            unique.\n\nchr_num\n\n            Number representing the chromosome to be processed.\n\nOPTIONAL INPUTS:\n\nreference_path\n\n            Path to the root subdirectory where the outputs of this\n            function are stored.\n            Default value: \"./genome/\"\n\nbamfile_path\n\n            Path to the subdirectory where the BAM file is located.\n            Default value: \"./indexedBAMfiles/\"\n\nmatrices_path\n\n            Path to the subdirectory where the output of this function\n            is stored.\n            Default value: \"./matrices/\"\n\npairedEnds\n\n            Flag for paired end read support. A value of true indicates\n            that the sequencer employed paired end reads, whereas a\n            value of false indicates that the sequencer employed single\n            end reads.\n            Default value: true\n\nnumBasesToTrim\n\n            A vector of integers specifying how many bases should be\n            trimmed from the begining of each read. If the vector\n            contains two integers, then the first integer specifies\n            how many bases to trim from the first read in a read pair,\n            whereas the second integer specifies how many bases should\n            be trimmed from the second read in the pair. If the\n            vector contains one integer, then all reads will have\n            that number of bases trimmed from the beginning of the\n            read. If no bases are to be trimmed, then this input\n            must be set to 0.\n            Default value: 0\n\nregionSize\n\n            The size of the genomic regions for which methylation\n            information is produced (in number of base pairs).\n            Default value: 3000\n\nminCpGsReqToModel\n\n            The minimum number of CpG sites within a genomic region\n            required to perform statistical estimation.\n            Default value: 10\n\nThe default values of regionSize and minCpGsReqToModel should only be  changed by an expert with a detailed understanding of the code and the  methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"estParamsForChr(mat_files,phenoName,matrices_path,reference_path,chr_num;\n                         estimation_path=\"./results/\",\n                         regionSize=3000,\n                         boundaryConditions=false)","category":"page"},{"location":"#InformMe.estParamsForChr-NTuple{5,Any}","page":"Home","title":"InformMe.estParamsForChr","text":"This function takes a list of BAM files (which correspond to the same  phenotype) and performs statistical model estimation within a specific  chromosome of interest. The function can be used on a computing cluster  to break the work of model estimation to many independent parallel job  processes. This is performed only after matrixFromBam.m.\n\nUSAGE (default):\n\nestParamsForChr(mat_files,prefix,matrices_path,reference_path,chr_num)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.\n\nestParamsForChr(mat_files,prefix,matrices_path,reference_path,chr_num,                  regionSize=2000)\n\nMANDATORY INPUTS:\n\nmat_files\n\n    All the .mat files to be included in the model. This can be a\n    single .mat file or multiple files in the form of a comma-sepa-\n    rated list of files.\n\nprefix\n\n    A string that specifies the name of the modeled phenotype.\n    The output files produced will contain this prefix.\n\nmatrices_path\n\n    A string that specifies the path to the directory that\n    where the output will be stored.\n\nreference_path\n\n    A string that specifies the path to the directory that\n    contains the results of analysis of the reference genome\n    performed by FastaToCpG.m as well as the results of\n    methylation calling performed by matrixFromBam.m.\n\nchr_num\n\n    Chromosome number 1 to 22 (in humans) specifying the\n    chromosome for which statistical estimation must be\n    performed.\n\nOPTIONAL INPUTS:\n\nregionSize\n\n    The size of the genomic region used for parameter\n    estimation (in number of base pairs).\n    Default value: 3000\n\nboundaryConditions\n\n    Flag to decide if boundary conditions should be estimated\n    freely in MLE.\n    Default value: false\n\nThe default value of regionSize should only be changed by an expert with  a detailed understanding of the code and the methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":" methAnalysisForChr(phenoName,chr_num,reference_path,estimation_path;\n                        outdir=\"./results/\",\n                        MSIflag=false,\n                        ESIflag=false,\n                        MCflag=false,\n                        regionSize=3000,\n                        subRegionSize=150)","category":"page"},{"location":"#InformMe.methAnalysisForChr-NTuple{4,Any}","page":"Home","title":"InformMe.methAnalysisForChr","text":"This function performs methylation analysis of a given chromosome in a  single phenotype. The function can be used on a computing cluster to  break the analysis work to many independent parallel job processes.  This is performed only after estParamsForChr.m in the Modeling  subdirectory is run to build the Ising models for the phenotype.\n\nUSAGE (default):\n\nmethAnalysisForChr(prefix,chr_num,reference_path,estimation_path)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.\n\nmethAnalysisForChr(prefix,chr_num,reference_path,estimation_path,                     outdir=\"/path/to/output\")\n\nMANDATORY INPUTS:\n\nprefix\n\n           A string that specifies the name of the phenotype to be\n           analyzed.\n\nchr_num\n\n           Chromosome number (1 to 22 in humans) specifying the\n           chromosome for which methylation analyis must be\n           performed.\n\nreference_path\n\n           A string that specifies the path to the directory that\n           contains the results of analysis of the reference genome\n           performed by FastaToCpG.m as well as the results of\n           methylation calling performed by matrixFromBam.jl.\n\nestimation_path\n\n           A string that specifies the path to the directory that\n           contains the results of parameter estimation performed\n           by estParamsForChr.jl.\n\nOPTIONAL INPUTS:\n\noutdir\n\n           A string that specifies the path of the directory in which\n           the methylation analysis results are written.\n           Default value: \"./results/\"\n\nMSIflag\n\n           Flag that determines whether this function performs\n           computation of the methylation sensitivity index (MSI).\n           false: no MSI computation.\n           true: allow MSI computation.\n           Default value: false\n\nESIflag\n\n           Flag that determines whether this function performs\n           computation of the entropic sensitivity index (ESI).\n           false: no ESI computation.\n           true: allow ESI computation.\n           Default value: false\n\nMCflag\n\n           Flag that determines whether this function performs\n           computation of turnover ratios, CpG entropies, capacities,\n           and relative dissipated energies of methylation\n           channels (MCs).\n           false: no MC computations.\n           true: allow MC computations.\n           Default value: false\n\nregionSize\n\n           The size of the genomic regions used for parameter\n           estimation (in number of base pairs).\n           Default value: 3000\n\nsubregionSize\n\n           The size of the subregions of a genomic region used\n           for methylation analysis (in number of base pairs).\n           The ratio regionSize/subregionSize must be an integer.\n           Default value: 150\n\nThe default values of regionSize and subregionSize should only be  changed by an expert with a detailed understanding of the code and  the methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"makeBedsForMethAnalysis(phenoName,analysis_path,reference_path;\n                              outdir=\"./singleMethAnalysisToBed_out/\",\n                              chrs=[string(\"chr\",i) for i=1:22],\n                              #minChrNum=1,\n                              #maxChrNum=22,\n                              MSIflag=false,\n                              ESIflag=false,\n                              MCflag=false,\n                              thresh=0.4,\n                              regionSize=3000,\n                              subregionSize=150)","category":"page"},{"location":"#InformMe.makeBedsForMethAnalysis-Tuple{Any,Any,Any}","page":"Home","title":"InformMe.makeBedsForMethAnalysis","text":"This function makes BED files for the methylation analysis results  obtained by means of MethAnalysisForChr.m for a single phenotype.\n\nUSAGE (default):\n\nmakeBedsForMethAnalysis(prefix,analysis_path,reference_path)\n\nUSAGE (optional):\n\nExample of optional usage with additional input parameters.\n\nmakeBedsForMethAnalysis(prefix,analysis_path,reference_path,                          outdir=\"/path/to/output\")\n\nMANDATORY INPUTS:\n\nprefix\n\n           A string that specifies the name of the phenotype.\n\nanalysis_path\n\n           A string that specifies the path of the directory in which\n           the model was constructed.\n\nreference_path\n\n           A string that specifies the path to the directory that\n           contains the results of analysis of the reference genome\n           performed by FastaToCpG.m as well as the results of\n           methylation calling performed by matrixFromBam.jl.\n\nOPTIONAL INPUTS:\n\nchrs\n\n          A vector of strings for the chromosomes to output to the\n          final bed files. Default value: `[string(\"chr\",i) for i=1:22]`\n\noutdir\n\n           A string that specifies the path of the directory in which\n           the output BED files are written.\n           Default value: \"./\"\n\nMSIflag\n\n           Flag that determines whether this function performs\n           computation of the methylation sensitivity index (MSI).\n           false: no MSI computation.\n           true: allow MSI computation.\n           Default value: false\n\nESIflag\n\n           Flag that determines whether this function performs\n           computation of the entropic sensitivity index (ESI).\n           false: no ESI computation.\n           true: allow ESI computation.\n           Default value: false\n\nMCflag\n\n           Flag that determines whether this function performs\n           computation of turnover ratios, CpG entropies, capacities,\n           and relative dissipated energies of methylation\n           channels (MCs).\n           false: no MC computations.\n           true: allow MC computations.\n           Default value: false\n\nthresh\n\n           A scalar used as a threshold in methylation-based\n           classification.\n           Default value: 0.4\n\nregionSize\n\n           The size of the genomic regions used for parameter\n           estimation (in number of base pairs).\n           Default value: 3000\n\nsubregionSize\n\n           The size of the subregions of a genomic region used\n           for methylation analysis (in number of base pairs).\n           The ratio regionSize/subregionSize must be an integer.\n           Default value: 150\n\nThe default values of thresh, regionSize, and subregionSize should only  be changed by an expert with a detailed understanding of the code and  the methods used.\n\n\n\n\n\n","category":"method"},{"location":"#","page":"Home","title":"Home","text":"diffMethAnalysisToBed(phenoName_1,phenoName_2,analysis_path_1,analysis_path_2,reference_path,\n                                  outdir=\"../makeBedsForDiffMethAnalysis_out/\",\n                                  chrs=[string(\"chr\",i) for i=1:22],\n                                  #  minChrNum=1,\n                                  #  maxChrNum=22,\n                                  MSIflag=false,\n                                  ESIflag=false,\n                                  MCflag=false,\n                                  regionSize=3000,\n                                  subregionSize=150,\n                                  minNumCpG=2,\n                                  thresh=0.55,\n                                  threshDMU=[-1,-0.55,-0.1,0.1,0.55,1],\n                                  threshDEU=[-1,-.5,-.3,-.05,.05,.3,.5,1])","category":"page"},{"location":"#InformMe.diffMethAnalysisToBed","page":"Home","title":"InformMe.diffMethAnalysisToBed","text":"This function makes BED files for the differential version of the  methylation analysis results obtained by means of MethAnalysisForChr.m  applied on two dinstict phenotypes.\n\nUSAGE (default):\n\nmakeBedsForDiffMethAnalysis(prefix_1,prefix_2,analysis_path_1,     analysis_path_2,reference_path)\n\nMANDATORY INPUTS:\n\nprefix_X\n\n           Strings with the first string specifying\n           the name of the first phenotype and the second string\n           specifying the name of the second phenotype used for\n           differential methylation analysis. Both phenotypes\n           must have already been analyzed with methAnalysisForChr.jl.\n\nanalysis_path_X\n\n           A string that specifies the path of the directory in which\n           the methylation analysis results obtained by\n           MethAnalysisForChr.jl are stored.\n\nreference_path\n\n           A string that specifies the path to the directory that\n           contains the results of analysis of the reference genome\n           performed by FastaToCpG.m as well as the results of\n           methylation calling performed by matrixFromBam.jl.\n\nOPTIONAL INPUTS:\n\nchrs\n\n           A vector of strings of chromosomes to be output to the final\n           bed files. Default value: `[string(\"chr\",i) for i=1:22]`\n\noutdir\n\n           A string that specifies the path of the directory in which\n           the output BED files are written.\n           Default value: \"./makeBedsForDiffMethAnalysis_out/\"\n\nMSIflag\n\n           Flag that determines whether this function performs\n           computation of the methylation sensitivity index (MSI).\n           false: no MSI computation.\n           true: allow MSI computation.\n           Default value: false\n\nESIflag\n\n           Flag that determines whether this function performs\n           computation of the entropic sensitivity index (ESI).\n           false: no ESI computation.\n           true: allow ESI computation.\n           Default value: false\n\nMCflag\n\n           Flag that determines whether this function performs\n           computation of turnover ratios, CpG entropies, capacities,\n           and relative dissipated energies of methylation\n           channels (MCs).\n           false: no MC computations.\n           true: allow MC computations.\n           Default value: false\n\nregionSize\n\n           The size of the genomic regions used for parameter\n           estimation (in number of base pairs).\n           Default value: 3000\n\nsubregionSize\n\n           The size of the subregions of a genomic region used\n           for methylation analysis (in number of base pairs).\n           The ratio regionSize/subregionSize must be an integer.\n           Default value: 150\n\nminNumCpG\n\n           The minimum number of CpG sites within an analysis\n           subregion required for performing full methylation-based\n           differential classification.\n           Default value: 2\n\nthresh\n\n           A scalar used as a threshold in methylation-based\n           differential classification.\n           Default value: 0.55\n\nthreshDMU\n\n           A 1x6 vector containing threshold values used for\n           methylation-based differential classification.\n           Default value: [-1,-0.55,-0.1,0.1,0.55,1]\n\nthreshDEU\n\n           A 1x8 vector containing threshold values used for\n           entropy-based differential classification.\n           Default value: [-1,-0.5,-0.3,-0.05,0.05,0.3,0.5,1]\n\nThe default values of regionSize, subregionSize, minNumCpG, thresh,  threshDMU, and threshDEU should only be changed by an expert with  a detailed understanding of the code and the methods used.\n\n\n\n\n\n","category":"function"},{"location":"#Index-1","page":"Home","title":"Index","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"","category":"page"}]
}
